source/src/drivers/cs_Uicr.c

```diff
@@ -0,0 +1,152 @@
+
+	// Based on the following post, but using nrf_nvmc functions where we can.
+	// https://devzone.nordicsemi.com/f/nordic-q-a/18199/dfu---updating-from-legacy-sdk-v11-0-0-bootloader-to-secure-sdk-v12-x-0-bootloader
+
+
+	// First block, contains all UICR->NRFFW[] and UICR->NRFHW.
+	const uint32_t startAddress = 0x10001014;
+	const uint32_t endAddress   = 0x10001080;
+	const int bufSize = (endAddress - startAddress) / sizeof(uint32_t);
+	uint32_t buffer[bufSize];
+
+	// Second block, contains all UICR fields after the UICR->CUSTOMER[].
+	const uint32_t startAddress2 = 0x10001200;
+	const uint32_t endAddress2   = 0x10001210;
+	const int bufSize2 = (endAddress2 - startAddress2) / sizeof(uint32_t);
+	uint32_t buffer2[bufSize];
+
+
+	CRITICAL_REGION_ENTER();
+
+	// Copy UICR to RAM.
+	uint32_t* address = (uint32_t*)startAddress;
+	for (int i = 0; i < bufSize; ++i) {
+		buffer[i] = *address;
+		address++;
+	}
+
+	address = (uint32_t*)startAddress2;
+	for (int i = 0; i < bufSize2; ++i) {
+		buffer2[i] = *address;
+		address++;
+	}
+
+    // Enable erase.
+    NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Een;
+    __ISB();
+    __DSB();
+
+    // Erase the page
+    NRF_NVMC->ERASEUICR = NVMC_ERASEUICR_ERASEUICR_Erase << NVMC_ERASEUICR_ERASEUICR_Pos;
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+
+    // Unlike nrf_nvmc_page_erase() we don't have to enable read here,
+    // as the next thing we do is go into write mode.
+
+	// Write the cached UICR contents back to the UICR.
+	nrf_nvmc_write_words(startAddress, buffer, bufSize);
+	nrf_nvmc_write_words(startAddress2, buffer2, bufSize2);
+
+	CRITICAL_REGION_EXIT();
+}
+
+cs_ret_code_t getUicr(cs_uicr_data_t* uicrData) {
+	uicrData->board                       = *((uint32_t*)g_HARDWARE_BOARD_ADDRESS);
+	uicrData->productRegionFamily.asInt   = *((uint32_t*)g_FAMILY_MARKET_TYPE_ADDRESS);
+	uicrData->majorMinorPatch.asInt       = *((uint32_t*)g_MAJOR_MINOR_PATCH_ADDRESS);
+	uicrData->productionDateHousing.asInt = *((uint32_t*)g_PROD_DATE_HOUSING_ADDRESS);
+	return ERR_SUCCESS;
+}
+
+cs_ret_code_t setUicr(const cs_uicr_data_t* uicrData, bool overwrite) {
+	// First check if every field can be written to UICR, we don't want a partial write.
+	if (!canSetUicrField(uicrData->board, g_HARDWARE_BOARD_ADDRESS)
+			|| !canSetUicrField(uicrData->productRegionFamily.asInt, g_FAMILY_MARKET_TYPE_ADDRESS)
+			|| !canSetUicrField(uicrData->majorMinorPatch.asInt, g_MAJOR_MINOR_PATCH_ADDRESS)
+			|| !canSetUicrField(uicrData->productionDateHousing.asInt, g_PROD_DATE_HOUSING_ADDRESS)) {
+		if (!overwrite) {
+			return ERR_ALREADY_EXISTS;
+		}
+		clearUicr();
+	}
+
+	// Write all fields.
+	setUicrField(uicrData->board, g_HARDWARE_BOARD_ADDRESS);
+	setUicrField(uicrData->productRegionFamily.asInt, g_FAMILY_MARKET_TYPE_ADDRESS);
+	setUicrField(uicrData->majorMinorPatch.asInt, g_MAJOR_MINOR_PATCH_ADDRESS);
+	setUicrField(uicrData->productionDateHousing.asInt, g_PROD_DATE_HOUSING_ADDRESS);
+	return ERR_SUCCESS;
```

Comment:
Assumes all nvm writes are successful. The nvm must be powered up. There shouldn't protection be enabled etc.

Reply:
What to do when it's not ready? Maybe we can add an assert (dev builds only)?

Reply:
Added checks, but no asserts yet.